src/node_messaging.cc

```diff
@@ -0,0 +1,829 @@
+    flag = args[2].As<Int32>()->Value();
+  }
+  Message msg(flag);
+  if (msg.Serialize(env, context, args[0], args[1])
+          .IsNothing())
+    return;
+  Send(std::move(msg));
+}
+
+void MessagePort::PostMessage(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  MessagePort* port;
+  ASSIGN_OR_RETURN_UNWRAP(&port, args.This());
+  if (!port->data_) {
+    env->ThrowError("Can not send data on closed MessagePort");
+    return;
+  }
+  port->Send(args);
+}
+
+void MessagePort::Start() {
+  Mutex::ScopedLock lock(data_->mutex_);
+  data_->started_ = true;
+  if (!data_->incoming_messages_.empty())
+    TriggerAsync();
+}
+
+void MessagePort::StartBinding(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  MessagePort* port;
+  ASSIGN_OR_RETURN_UNWRAP(&port, args.This());
+  if (!port->data_) {
+    env->ThrowError("Cannot start listening on a closed MessagePort");
+    return;
+  }
+  port->Start();
+}
+
+void MessagePort::MoveToContext(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  MessagePort* port;
+  if (!args[0]->IsObject() ||
+      (port = Unwrap<MessagePort>(args[0].As<Object>())) == nullptr) {
+    env->ThrowTypeError("First argument needs to be a MessagePort instance");
+  }
+  if (!port->data_) {
+    env->ThrowError("Cannot transfer a closed MessagePort");
+    return;
+  }
+  if (port->is_privileged_ || port->fm_listener_) {
+    env->ThrowError("Cannot transfer MessagePort with special semantics");
+    return;
+  }
+  Local<Value> context_arg = args[1];
+  Local<Context> context;
+  if (!context_arg->IsObject() ||
+      !ContextFromContextifiedSandbox(env, context_arg.As<Object>())
+          .ToLocal(&context)) {
+    env->ThrowError("Invalid context argument");
+    return;
+  }
+  Context::Scope context_scope(context);
+  MessagePort* target =
+      MessagePort::New(env, context, nullptr, std::move(port->data_));
+  if (target) {
+    args.GetReturnValue().Set(target->object());
+  }
+}
+
+size_t MessagePort::self_size() const {
+  Mutex::ScopedLock lock(data_->mutex_);
+  size_t sz = sizeof(*this) + sizeof(*data_);
+  for (const Message& msg : data_->incoming_messages_)
+    sz += sizeof(msg) + msg.main_message_buf_.len;
+  return sz;
+}
+
+void MessagePort::Entangle(MessagePort* a, MessagePort* b) {
+  Entangle(a, b->data_.get());
+}
+
+void MessagePort::Entangle(MessagePort* a, MessagePortData* b) {
+  MessagePortData::Entangle(a->data_.get(), b);
+}
+
+MaybeLocal<Function> GetMessagePortConstructor(
+    Environment* env, Local<Context> context) {
+  // Factor generating the MessagePort JS constructor into its own piece
+  // of code, because it is needed early on in the child environment setup.
+  Local<FunctionTemplate> templ;
+  templ = env->message_port_constructor_template();
+  if (!templ.IsEmpty())
+    return templ->GetFunction(context);
```

Comment:
Is `GetFunction()` cached by context?

Reply:
Yup – I have to admit I never got the point of `{Function,Object}Template`s before making this PR. However, in retrospect it’s obvious: It’s a `Template` in the sense that it can be used to create functionally equivalent functions/objects in different contexts.